crates/core/src/nucleus/validation/build_from_dht.rs

```diff
@@ -0,0 +1,248 @@
+use crate::{
+    context::Context, entry::CanPublish, network::entry_with_header::EntryWithHeader,
+    workflows::get_entry_result::get_entry_with_meta_workflow,
+};
+use holochain_core_types::{
+    chain_header::ChainHeader,
+    entry::{Entry, EntryWithMeta, EntryWithMetaAndHeader},
+    error::HolochainError,
+    time::Timeout,
+    validation::{ValidationPackage, ValidationPackageDefinition},
+};
+
+use std::sync::Arc;
+
+const GET_TIMEOUT_MS: usize = 1000;
+
+async fn all_chain_headers_before_header_dht(
+    context: Arc<Context>,
+    header: &ChainHeader,
+) -> Result<Vec<ChainHeader>, HolochainError> {
+    let mut current_header = header.clone();
+    let mut headers = Vec::new();
+
+    while let Some(next_header_addr) = current_header.link() {
+        log_debug!(context, "About to try and get header: {}", next_header_addr);
```

Comment:
If ChainHeader had a method to follow the link by link chain we could be able to to fold this without changing current_header state?

current_header.follow_links().fold(Ok(Vec::new()),|headers,chain_header|{})